src/features/broker-protection/comparisons/is-same-name.js

```diff
@@ -20,16 +20,26 @@ export function isSameName (fullNameExtracted, userFirstName, userMiddleName, us
     userLastName = userLastName ? userLastName.toLowerCase() : null
     userSuffix = userSuffix ? userSuffix.toLowerCase() : null
 
-    // check their nicknames too
-    const nicknames = getNicknames(userFirstName)
+    // Get a list of the user's name and nicknames / full names
+    const names = getNames(userFirstName)
 
-    for (const firstName of nicknames) {
+    for (const firstName of names) {
     // Let's check if the name matches right off the bat
         const nameCombo1 = `${firstName} ${userLastName}`
         if (fullNameExtracted === nameCombo1) {
             return true
         }
 
+        // If the user didn't supply a middle name, then try to match names extracted names that
+        // might include a middle name.
+        if (!userMiddleName) {
+            const matchesFirstAndLast = new RegExp(`^${firstName}\\s?.*\\s${userLastName}$`)
+
+            if (matchesFirstAndLast.test(fullNameExtracted)) {
+                return true
+            }
```

Comment:
Let's make this driven by array comparisons instead of regex. Mostly because I think regex should be a last resort, but more importantly because the next iteration of matching names (and everything else) can be driven in this data-first, declarative style too (instead of how it's implemented now with string interpolation etc)

so, to remove this regex here + be inline with future changes, we could do something like this:

```suggestion
            const elements = fullNameExtracted.split(' ')
            // create the subject, in this case first + last elements in the string
            const subject = [elements[0], elements[elements.length - 1]]
            // target is current name/nickname + user entered last name
            const target = [firstName, userLastName]
            // compare
            const matchesFirstAndLast = subject.join(' ') === target.join(' ')

            if (matchesFirstAndLast) {
                return true
            }
```

Reply:
@shakyShane I considered doing this in my first approach, but I'm worried that this won't reliably get the first and last name (for instance, when someone has a last name with a space in it), which is why I went the regex route. 

Since the full extracted name string is the source of truth, it seemed best to try and not mess with the formatting too much since we don't know exactly what we're going to get. What do you think?

Reply:
This conversation is proof we don't have enough test data yet - because if either of us thinks an approach is 'better', then we should be able to encode that proof into the tests. 

so yeah, if you wanted to keep the regex (I appreciate how you have this problem in your head better than I do, so I trust your judgement), could we at least add another test case that would fail with my example? that way we're building up more verifications for when we refactor

Reply:
Yeah, that's fair enough. I will do. I think we'll have enough test data soon to switch approaches (or alter the proposed approach to fit).

Reply:
I think the regex approach is fine since we want to tend to overmatch rather than undermatch. Agree we should have a "false positive" test case to make sure we don't go too far in the other direction.

Reply:
with the following test, you can see that interpolating user data into regexes is overly risky

```js
it('property testing isSameName (seed 1)', () => {
    // this seeded example showed how we needed to escape the user-values in the regex
    fc.assert(fc.property(
        fc.string(),
        fc.string(),
        fc.option(fc.string()),
        fc.string(),
        fc.option(fc.string()),
        (fullNameExtracted, userFirstName, userMiddleName, userLastName, userSuffix) => {
            const result = isSameName(fullNameExtracted, userFirstName, userMiddleName, userLastName, userSuffix)
            expect(typeof result).toBe('boolean')
        }
    ), { seed: 493629956, path: '50:1:0:0:1:2:1:1:6:7:7', endOnFailure: true })
})
```

with that seed, you get:

<img width="575" alt="Screenshot 2024-02-29 at 1 25 50 PM" src="https://github.com/duckduckgo/content-scope-scripts/assets/1643522/a6318395-7c31-45b2-abce-e19d90b6ef26">

Looking again at the problem, we're just saying:
- a string that begins with `x` 
- AND ends in `y`
- AND has letters/spaces inbetween

therefor, this would work

```js
        if (!userMiddleName) {
            const combinedLength = firstName.length + userLastName.length
            const matchesFirstAndLast = fullNameExtracted.startsWith(firstName) &&
                fullNameExtracted.endsWith(userLastName) &&
                fullNameExtracted.length > combinedLength
            if (matchesFirstAndLast) {
                return true
            }
        }
```

Reply:
I’m happy with that approach. 👍